RFC - Type extension and duplication (#428)
blockprotocol/blockprotocol: rfcs/text/0428-extending-and-duplicating-types.md

It is important to note that aside from difficulties in implementations, the actual semantics of a type with a cyclical inheritance tree are well defined. As the types largely follow JSON schema semantics, constraints are composed and never overridden. Any finite cycle of types would result in a finite set of constraints, and upon detecting a cycle, an implementation can safely finish traversal once establishing the comprehensive set of constraints (therefore avoiding an infinite loop).

## Disallowing unsatisfiable schemas

Another uncomfortable side-effect of the extension mechanism is that it gives more routes through which an 'unsatisfiable type' can be created. An _unsatisfiable_ type is a type where it is impossible to create data that satisfies the constraints of the type.

This can happen through a combination of constraints which overlap or contradict one another, where a simple example is saying a property must be both a number and an array.

> Prior to this proposal, the primitives over which the type system are built were limited enough to avoid being able to define conflicting constraints. However, with the introduction of type extension, it will be possible to combine two schemas which express a constraint over the same thing, and there is no guarantee that those constraints will be compatible.

Reviewer: Applications can prevent this with their own handling logic - for example by validating that when updating the version of a type one is inheriting from, it is checked for conflicts - correct? This is simply not spec-guaranteed, if I understand correctly. @Alfred-Mountfield @TimDiekmann

Reviewer: Nevermind, I read on.

Reviewer 2: Yes. We don't enforce valid constraints as this would be very hard (if possible at all) to encapsulate in the type system itself. Applications are responsible to validate the entity types - or at least they should. If they don't the worst thing which could happen is that an entity of this type can not be created.

#### Unequal and incompatible overlapping constraints

Finally, say that `Person` defines that `name` is a _string_ and is **required**, and `Superhero` defines that `name` is _an array of strings_ and is **required**.

When applied simultaneously, these constraints are _incompatible_ and result in an unsatisfiable type. Let us try and create an instance that satisfies it. First we try and create one with a string `name`:

```json5
{
  name: "John Doe", // ERROR: expected string array, got string
  ... // superpower, age, occupation
}
```

And then we instead try and create one with an array of `name`s:

```json5
{
  name: ["John Doe"], // ERROR: expected string, got string array
  ... // superpower, age, occupation
}
```

### An uncomfortable compromise

As alluded to above, checking compatibility of any two given schemas is quite possibly an NP-Hard problem. Even if it is generally solvable, it is likely a task that is complicated enough that it would be an exceptionally high barrier to entry for potential implementers to overcome.

> While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we opt not to encode such restrictions within the type system, and **implementations should not assume that an arbitrary type is satisfiable**.

Reviewer: ```suggestion
While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we are not proposing (at this time) to encode such restrictions within the type system itself, and implementations therefore SHOULD NOT assume that an arbitrary type is satisfiable.
```

- Feature Name: extending-and-duplicating-types
- Start Date: 2022-07-11
- RFC PR: [blockprotocol/blockprotocol#428](https://github.com/blockprotocol/blockprotocol/pull/428)
- Block Protocol Discussion: [blockprotocol/blockprotocol#439](https://github.com/blockprotocol/blockprotocol/discussions/439)

# Summary

[summary]: #summary

> As a follow-up to the [Graph type system RFC](./0352-graph-type-system.md), this RFC will describe the behavior of which types in the type system can be extended and duplicated to enhance reusability while giving more control to users (both block authors and users of embedding applications). Extending types can be considered the same as "type inheritance", but there are some important nuances that make these concepts different.

Reviewer 3: This has been reworked

The types in the type systems assume an implicit `{ "additionalProperties": false }` JSON Schema attribute that has implications for how the schemas can be composed together. The presence of this option will be addressed in this proposal.

Allowing entity types to be extended in the Block Protocol means that a user could still make use of public types when they want to define types for their domain. As an alternative to extending types, type duplication (or type forking) is a less compatible way to make use of existing public types, without a direct data mapping strategy.

This RFC introduces a way for types to be extended in a way where the reusability and sharing design philosophies of the Block Protocol are maintained, and a way for types to be duplicated for use cases where extending a type is insufficient.

# Guide-level explanation

[guide-level-explanation]: #guide-level-explanation

> > ⭐ Usage of the word "types" in this RFC can be assumed to refer to _entity types_ unless stated otherwise.

Reviewer 3: We've reworked it to be clearer that this is about entity types. We can revisit for the others if needs be at some point

The assumption that we can select/project parts of a subtype that make up a supertype is essential for keeping strictness in JSON Schemas.

We propose slight modifications to how `{ "additionalProperties": true }` and `{ "unevaluatedProperties": false }` behave and may be used within our type extension system to let supertypes keep strictness while allowing composition. This modification is a semantic change and does not change any syntax.

Concrete examples of how JSON Schema breaks with these validation constraints are shown in the [Reference-level explanation](#problems-with-unevaluatedproperties).

## Defining extended entity types

Extended types will be defined with conventional JSON Schema syntax: the `allOf` keyword. An entity type can extend another entity type by adding a versioned URL reference to the root-level `allOf` array.
A [versioned URL](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0408-versioning-types.md#type-uris) is used so that subtypes aren't automatically updated (and potentially invalidated) when the supertype is updated.

> As extended types can extend other extended types we must also make sure that there are no cycles within the type hierarchy, as it makes types difficult to resolve/reason about and could lead to unpredictable behavior.

Reviewer 3: We've removed this, and explained it within the rationale section

## Defining entity type duplication

When looking for public types, if extending an entity type is insufficient, an alternative is to duplicate the type so that individual properties can be removed or overridden. Duplicated types become new complete, standalone types.

Type duplication, unlike extended types, does not imply that instances of the original type can be substituted by instances of the duplicated type because the types may be incompatible. A mapping may be used to persist some of the compatibility, but mapping is yet to be defined within the type system.

> ### Duplication strategy

Reviewer 3: This has been reworked in the new draft of the document, will leave the comment thread open for visibility.

{
  "city": "Copenhagen",
  "name": "Charles"
}
```

results in validations errors such as `Property 'name' has not been successfully evaluated and the schema does not allow unevaluated properties`.

Changing from `unevaluatedProperties` to `additionalProperties` results in errors `Property 'name' has not been defined and the schema does not allow additional properties`.

Both of these solutions for strict schemas would not be suitable for the type of expressiveness we want for type extension, unfortunately.

The required behavior is that `unevaluatedProperties` should only validate at the top level of a type, allowing supertypes to validate `unevaluatedProperties` but defer checking if they're used within in subtype.

### Redefining `unevaluatedProperties`

We're already implicitly defining `{ "additionalProperties": false }` in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) for all schemas, this RFC will piggyback on that, and change the existing, implicit `additionalProperties` usage to `unevaluatedProperties` and slightly change its [vocabulary definition](https://json-schema.org/draft/2020-12/json-schema-core.html#name-unevaluatedproperties).

> The vocabulary change is that the `unevaluatedProperties` keyword will only be applicable at top-level schemas (i.e. subtypes or schemas that don't specify `allOf`), and ignored when present on a schema that resides in the `allOf` array. Instead of the default value being `{ "unevaluatedProperties": {} }`, the default value will be `{ "unevaluatedProperties": false }`. Together this makes it so that all supertypes will be able to compose, as their individual, implicit `unevaluatedProperties` values won't have any effect on validation and defer evaluation to the extending type (the subtype).

Reviewer 3: I think it would be better to not change the vocabulary of JSON schema and instead just say we _implicitly introduce_ `unevaluatedProperties` to the top level of an entity type when we consider it for validation.

(I think we should do this _if_ we don't split the schemas into constituent parts as suggested for consideration later on)

Reviewer 3: This isn't that different to what's written here, but as long as we just don't include `unevaluatedProperties` within the meta-schemas, it won't be possible (valid) to introduce the keyword in other places, and we don't need to change the vocabulary to say we "ignore it" or otherwise.

Reviewer 3: This has been written into the document and tested locally

First entity type version of `Country` without supertype:

```json
{
  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
  "kind": "entityType",
  "$id": "https://blockprotocol.org/@alice/entity-type/country/v/1",
  "type": "object",
  "title": "Country",
  "properties": {
    "https://blockprotocol.org/@alice/property-type/name": {
      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
    }
  },
  "required": ["https://blockprotocol.org/@alice/property-type/name"]
}
```

Entity type `Region` with `Country` as a supertype

```json
{
  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
  "kind": "entityType",
  "$id": "https://blockprotocol.org/@alice/entity-type/region/v/1",
  "type": "object",
  "title": "Region",
  "allOf": [
    { "$ref": "https://blockprotocol.org/@alice/entity-type/country/v/1" }
  ],
  "properties": {
    "https://blockprotocol.org/@alice/property-type/blurb": {

>       "$ref": "https://blockprotocol.org/@alice/property-type/blurb/v/1"

Reviewer 3: Not important, but did we mean to have blurb here?

    }
  },
  "required": ["https://blockprotocol.org/@alice/property-type/blurb"]
}
```

The second, cyclic version of the `Country` entity type

```json
{
  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
  "kind": "entityType",

>   // Because of versioning, we cannot change this version to /v/1 and create a "proper" cycle.

Reviewer 3: Is this true? 

The schemas are just plain records, we don't have any concept of the times they were created at. It would be possible to publish two types that depend on each other in the same version. I _could_ write a type which depends on itself in version 1.

    },
    "https://blockprotocol.org/@alice/property-type/location": {
      "$ref": "https://blockprotocol.org/@alice/property-type/location/v/1"
    }
  },
  "required": ["https://blockprotocol.org/@alice/property-type/name"]
}
```

This sort of type hierarchy should _not_ be encouraged within the type extension system, as the circular dependencies make types difficult to reason about. While the type hierarchy might be completely valid (as it would be in this case), we should safeguard users from making redundant type structures that look like the above.

In this specific contrived example, creating a new entity type based on `Region` instead of a new version of `Country` might even encode semantic meaning better than re-defining `Country`.

## Multiple supertypes - checking compatibility

As described in the [Guide-level explanation](#multiple-supertypes), when extending multiple entity types, they must be able to coexist in a "compatible manner", which means that the entity types' `properties` and `links` comply with the following:

- For each property (base URL on the top level of `properties`) that exists in multiple entity types:

>   - all entity types refer to the same versioned URL of the property type **or** compatible versions of the property type (through the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) definition of 'compatible')

Reviewer 3: This is perhaps the biggest open question to me now. I wonder if structural compatibility is good enough, maybe it is.

I have concerns, especially in the current system where we don't have fine-grained constraints on values, that the semantic meaning of property types is important. What if I had a "Column ID" field in one of my entity types that I use in my API, and it's simple a `Text` value, with a `description` that says "must be the ID of a specific account". And then what happens if a new version of "Column ID" is published which modifies the description to be "must be the ID of a specific account within Europe".

This is perhaps one of the compromises we have to be happy with, because not everything will be able to be rigorously encoded, and there needs to be some leeway whereby things can be updated with ease. Relying on everything having the same version of a type sounds like a maintenance nightmare if you don't manage all the types, and you might be waiting on some upstream creator to update their type to have the new version of `Name` which fixed a typo in the description.. which would just encourage forking.

Reviewer 3: The requirements about compatibility have been completely removed in the new version of the doc. The relevant section is in the rationale and alternatives, and describes the difficulty of enforcing satisfiable schemas.

  - none of the entity types define the property type as an array **or** all define the property type as an array with compatible cardinality constraints.

- For each link (versioned URL on the top level of `links`) that exists in multiple entity types:

>   - none of the entity types define the link as an array **or** all define the link as an array with compatible cardinality constraints (and same `order` value).

Reviewer 3: All links are implicitly captured as "arrays" now. We just need to enforce compatible cardinality constraints, **destination** constraints, and ordering.

Compatible destination constraints is less clear to me.. We should be specific about this, I'm guessing we have to check compatibility of the union of link destinations that are allowed, where compatibility between entity types is effectively just accounting for inheritance? Or do we also need to consider versioning here 🙃

Reviewer 3: This is now redundant in the reworked document as we don't require type extension to result in something satisfiable.

Using the above setup would mean that we need to specify `#open` at the end of type URLs when URLs appear in `allOf`. We would also need to serve type schemas with one of the above structures, such that they conform with JSON Schema. Although this does add some ceremony around extending types, we would end up with a type extension system that would conform to JSON Schema without implicitness or redefining keyword semantics.

# Prior art

[prior-art]: #prior-art

- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)

# Unresolved questions

[unresolved-questions]: #unresolved-questions

- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.

> - The current argument for not allowing cyclic type hierarchies mostly build on a feeling that type hierarchies shouldn't be too indirect/obfuscated, but there could be stronger arguments for allowing/disallowing it.

Reviewer 3: ➕

> - The decision to have type substitution between subtype and supertype does mean the extended types are restrictive in that they do not allow for removing/changing existing properties. Without a way to transform data between types (mappings), allowing overrides is not possible while keeping compatibility.

Reviewer 3: I think we mention mappings too many times in this RFC without explaining what "mappings" are. 

They're currently an internal only thought, and we don't have a good resource to point to, so this would be a good opportunity to raise the general idea of them, and we can look at publishing a draft placeholder RFC to create something referable soon.

Reviewer: Happy to help with this.

As the new Block Protocol type system doesn't require a delta-based storage approach for its schemas, it's unclear what the advantages of partial schemas for update requests are.

Furthermore, resolving the intentions of the existing definition for the message is problematic, as it's impossible to know whether the omission of a property should be interpreted as _removing_ that property, or leaving it in place.

We propose that we should treat type updates as complete replacements, so implementation is much more straightforward for embedding applications. Partial schema updates also add some level of indirection, and that may obfuscate error sources and error reasons. Therefore, updates to types must be complete replacements rather than partial schema updates.

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## Defining extended entity types

> In the Block Protocol, we will allow type extension through the `allOf` JSON Schema keyword which applies all constraints from the sub-schemas within the `allOf` array..

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/1298

Any instance of `Employee` will have a `name` property, which will always satisfy the constraints of `name` on `Person`, as being required is a _stronger_ constraint than being optional, and all other constraints (imposed by the property type) are equal.

### Problems with `additionalProperties` and `unevaluatedProperties`

The intention of the type system is to strongly type the data, so that users of the system can benefit from strong guarantees about the shape of, and existence (or non-existence) of parts of the data.
As such, entity types are intended to comprehensively describe data within entities of that type, which means we wish to disallow additional or untyped properties within them.

In JSON schema this is generally achieved by setting `additionalProperties: false`, or `unevaluatedProperties: false`.
Unfortunately, this causes problems when you start composing schemas (e.g. with entity type inheritance)

If supertypes themselves specify one of these `{ "unevaluatedProperties": false }`, they are not able to be part of an `allOf` validator, as they will error as soon as they see properties that are not part of the root schema itself.

> When composing schemas that all contain `{ "unevaluatedProperties": false }`, each schema will disallow any other properties which they do not define. Using the following JSON Schema as an example:

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/1298

While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we opt not to encode such restrictions within the type system, and **implementations should not assume that an arbitrary type is satisfiable**.

# Prior art

[prior-art]: #prior-art

- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)

# Unresolved questions

[unresolved-questions]: #unresolved-questions

- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.

>   - This may be especially difficult if a property type is defined as being an array of `oneOf` a given set of property type objects. Detecting which sub-schema applies to a given array element when accounting for dropped properties, **may** be a very difficult task.

Reviewer 2: I don't really get what you mean by "projecting/selecting". If `Employee` is a subtype of `Person` and for example and `Person` has a `oneOf` property, what is the difficulty to use `Employee`? Let's assume an application, which requires a `Person`. If an `Employee` is passed to the application, the type still satisfies `Person` so the application can _just use_ the provided type. The application should only rely on the `Person`'s properties to exist.
If the `oneOf` is an array, either all properties are valid as the array is of a specific type, which satisfies `Person` or the types are not compatible.

I don't think, this is an issue.

Reviewer 2: ```suggestion
```

Reviewer 3: > the type still satisfies Person so the application can just use the provided type

This is the assumption I think we're disagreeing on. We're not only interested in duck-typing (or similar patterns). Entity types behave different to a `type` or `interface` in TypeScript, they do not say "the object has to _at least_ have X" they define the **complete** set of possible entities.

This is the point above about needing a "projection" or "selection" of properties. If a block asks for a `Person` and you pass in `Employee`, that employee entity will _not_ satisfy the `Person` schema because we do not allow `additionalProperties: true` (or `unevaluatedProperties`, hence the section on that).

This has even more run on affects we need to consider, such as the `entityTypeId` of a 'projected' entity. The main issues though I think come about from use-cases that deal with `create` or `update` methods for entities. If they receive a subtype of the other type and `update` it, but are only aware of _some_ of the properties, then they might accidentally invalidate the subtype. For an example, let's think you have a component that lets you duplicate a `Person` entity (not to be confused with "duplication" of types outlined in this post). The component might let you pick from a list of `Person` entities, pick one to copy the properties over from, and then edit it. This list is going to contain `Employee` entities as well, as they're a subtype of `Person`. But now you try and copy the properties of one of the `Employee` entities. If you literally did `deepcopy(person.properties)` you'd get the additional properties from `Employee` and it'd break validation when you try and `createPerson`.

Reviewer 4: > If we allow setting additionalProperties in entity types, it's possible to mark an entity as "final" - an entity type that cannot be inherited from. This would be possible by adding additionalProperties: { type: boolean } to properties on the entity type meta schema. As we currently do not enforce additionalProperties: false on entity types I think this is good for allowing applications to treat the entity types as they need to.

How exactly would that be reflected in the graph itself? I get your POV, but from my perspective, this is full of potential problems when trying to build on-top of these assumptions.

What I mean is the following: if I have a type `Person` with properties: `firstName`, and `lastName`, if we're going to assume `additionalProperties: true`, then it's valid to also have `middleName` in there (also when creating the type). What will happen in the case of, e.g. migration tools if someone now wants to add `middleName` as a "real" property?

I think this is then also ripe for abuse, where people implicitly expect (even if not in the schema) that a specific property exists (because the developer just adds it to their entities and then uses it in their blocks).

Reviewer 2: We assume in the version-RFC an implicit `additionalProperties: false` to ensure, that entities do not have random properties, but this does not work with subtyping (as mentioned by @Alfred-Mountfield above)

Reviewer 3: > The same applies to create and update: If the application does not correctly creates or updates an entity it's the application's fault.

I had an example that I wanted to walkthrough to discuss this, but I think it's a distraction, I think the discussion here is actually about the need for a projection mechanism.

We are discussing the need for a 'projection' here to actually satisfy the subtype property of our inheritance mechanism. It is important that a block is able to request a `Person`, and then be able to validate it against that `Person` type. If it is of type `Employee`, we want to make sure the additional properties in it are adequately captured by `Employee`, we don't want to _**also**_ allow any other random property like "potato: 32" which isn't even associated with a property type.

Reviewer 4: @Alfred-Mountfield has perfectly summarized what I was trying to say about `additionalProperties`.

I'd suggest that instead of "we suggest the constraint is implied at the root level schema when validating data against a specific entity type" we strengthen the wording to: "the constraint **must** be implied at the root level schema when validating data against a specific entity type". Otherwise, it might lead to a split in the ecosystem if some implementation decide not to do so.

Reviewer 2: > Is literally describing duck-typing. This is how TypeScript interfaces and types work.

I assumed another location where the duck typing happens. You don’t need to duck-type on your application as you expect the entity to be a specific entity type. Validating a JSON based on a given JSON schema is always duck typing. We don't expect the blocks to check, if `name` exists on a `Person`, it can assume it's there, but the application may need to make sure that only `Person`s are passed to the blocks.

---

However, I understand your point now and we indeed talked passed each other. As to me, it was not clear that by "projecting" you mean "coercion", which is probably the more common term when talking about subtyping and variance. For general JSON data to be passed as subtype, coercion _should_ be fairly easy but as outlined in the code it's not defined how this would happen

To conclude this thread: Do you think it's required to resolve this question prior to merging the RFC?

- Type extension might prove to have strange side-effects when considering link types which extend from one another.

>   - Say `Person` defines `Knows` and `Has Friend` links. If `Has Friend` extends `Knows`, then technically any `Has Friend` instance is also a `Knows` instance, and would need to satisfy the link constraints set on both. This has implications for `minItems` and `maxItems`, as well as for link destinations. If `Person` says it only `Knows` other `Person` entities, but `Has Friend` links to both `Person` and `Pet` entities, then a `Has Friend` link to a `Pet` entity would break the constraint set for `Knows`.

Reviewer 3: I'm unsure if this is a **problem** yet.

It could be quite similar to the situations outlined above where the constraints are just combined, and in some cases that leads to an unsatisfiable condition (whereby the user then needs to update their type to make it satisfiable).

Reviewer: Right. I think that's the appropriate resolution strategy. If people's definitions break logical transitivity principles we can prompt them to update those assumptions (updating types themselves, or forking/changing types).

Reviewer 2: I was about to comment on this exact line prior to seeing your comment.
My instinct is that this is **not** a problem as the worst thing which could happen is an unsatisfiable constraint, which we explicitly opted out of.

```suggestion
- Type extension might prove to have strange side-effects when considering link types which extend from one another.
```

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/1298

>   - Say `Person` defines `Knows` and `Has Friend` links, and that those links are **ordered**. At the moment, the order of a link is defined in a given direction (e.g. left to right) and has a single value. That would presently imply that the order of the `Knows` link _within the `Knows` links list_ would have to be the same as the order of the `Has Friend` link _within the `Has Friend` links list_.

Reviewer 3: This is definitely a problem, and probably blocks acceptance. 

One potential route is to somehow 'bucket' the lists of links, and have an order per list.

Another potential route is to rework our ordering strategy where we define an `orderBy` mechanism on the source entity type, rather than having a native definition of `order` _within_ link entities. Something like this was proposed by @indietyp a while ago, and has sat in the background until we needed to revisit link ordering (thinking about race-conditions, fractional indexing, consistency guarantees, etc.), now is probably the time where we need to revisit it.

cc: @CiaranMn

Reviewer: OrderBy would work, if "user-specified custom" were an option it allowed.

But that option needing to exist would then again necessitate a solution like buckets.

Reviewer 3: > OrderBy would work, if "user-specified custom" were an option it allowed.

Yes I guess the main problem of something like `OrderBy` is it precludes people from designing something like a Spotify playlist. You could easily do things like alphabetical or chronological ordering (which you can't do currently easily), but allowing people to reorder and specify a defined ordering.. that would be local to the playlist, and would have the same issue outlined above about a link appearing under multiple 'collections'.

A potential route away from this, is that we provide an `orderBy` mechanism on links but then also establish a second pattern wherein entities store local definitions of orderings within their properties. So, a `Playlist` entity could store an optional `Song Ordering` property, which is an array of link entity IDs. I _**really**_ don't like this though, and the main reasoning for that is explained within #1020 and is to do with "referability" and consistency guarantees. If a different block / user / whatever were to delete one of the link entities, or somehow make them invalid, the `Playlist` entity would _not_ be implicitly updated because there would be nothing within the type system to enforce it (like there _**is**_ with actual links), and this causes a massive risk vector for race conditions.

On the other hand, one of the issues with `OrderBy` (and why we didn't continue exploring it in detail), is that really that's something a user can do at point of querying, `queryEntities` for instance already allows you to define an `OrderBy` constraint, which could affect pagination. The main benefit would be allowing backends to optimise their storage layout for ordering in the first place and have faster queries, but considering it doesn't solve the core functionality problems here.. we still need to think about the `Playlist` case.

> necessitate a solution like buckets.

I'm hesistant to argue for buckets / collections, but I've brought it up offhand in conversations multiple times over the past few months so it seems like it keeps coming up as a prominent 'obvious' solution to certain things, and maybe it's time to accept it as worth the investment. The investment isn't minor, existing datastores would have to quite heavily change how they store links now (somehow grouping them together which is not a concept we currently have at all), the type system and its implementations would need updating, the query mechanisms would need to change, the `Subgraph` would somehow need to have a way of grouping link entities (tying them to the source entity type definition too which is a bit strange), the standard library would need to change..

Hm

Reviewer 4: I have 3 distinct ideas, that have been bouncing around my head over the last couple of hours. Thought I'd contribute them to the discussion. :D

## link type `orderBy`: Value

We reserve a specific property type (e.g. `blockprotocol/@blockprotocol/orderBy`), if that property type is present in the definition it is ordered. This is not ideal, as we would lose quite a lot of flexibility.

## link type `orderBy`: Path

What if `orderBy` would be defined on the link type itself, but instead of being a value, it would be a (JSON)-pointer to an existing property on the object itself.

The property needs to be a comparable value (preferably a `Number`). The interesting part about this is that it allows the user to choose which property to use freely and can be (pretty efficiently) stored in data stores (by just computing the value).

What would be more complicated is to allow traversal via the JSON pointer (kind of like what we're currently doing with the)

## entity type `orderBy`: Path

instead of defining a specific value, we define a path to a property, the property it is pointing to needs to be of a valid datatype (preferably `Number`). This is very similar to the second proposal, except that we're now able to query the link `rightEntity` (and `leftEntity`) safely, if we navigate to the `rightEntity`, we can only use properties to order that are present on all valid destinations.

### Example

```
EntityType Playlist {
	properties = {
		title,
		description
	}

	links = {
		includedInPlaylist = {
			orderBy = ["properties", "position"]
			// as an alternative
			orderBy = ["rightEntity", "properties", "chartPosition"],
			oneOf = [Song]
		}
	}
}

EntityType Song {
	properties = {
		title,
		description,
		chartPosition,
	}
}

EntityType includedInPlaylist {
	oneOf = [link],

    properties = {
		position
	}
}
```

(the syntax is completely made up, needed something terse as I didn't want to post a whole schema)

What is interesting about this is that by just changing the ordering in `Playlist` we would be able to create custom "auto-generated" playlist positions, that is depending on either the `Song` or the "local" position.

(Of course, the type system would need to have the ability to re-define the `orderBy` attribute)

Reviewer 4: Thanks! To get us all on the same page, I have constructed a (realistic) example that would make exhibit the problem,

let's say we have the following entity types:
* `Person`
* `Address`
* `WorkAddress`
* `HomeAddress`
* `HomeOfficeAddress`
* `Location`

`Person` has 4 links:
* `Person --Address-> Location (orderBy: preference)`
* `Person --WorkAddress-> Location (orderBy: preference)`
* `Person --HomeAddress-> Location (orderBy: preference)`
* `Person --HomeOfficeAddress-> Location (orderBy: preference)`

`Address` defines a property: `preference`, and we have the following inheritance scheme:

```
        Address
       /       \
WorkAddress    HomeAddress
       \       /
    HomeOfficeAddress
```

As an example let's say we have as a user `Max Mustermann`, `Max Mustermann` being the rich banker that he is, is able to afford multiple homes and work locations, doing so we have:

```mermaid
flowchart LR
	mm[Max Mustermann] -->|WorkAddress - 0| Frankfurt
    mm -->|WorkAddress - 10| Berlin
    mm -->|HomeAddress - 20| Hochheim
    mm -->|HomeAddress - 30| Dortmund
    mm -->|HomeOfficeAddress - 40| Ibiza
```

We can generalize these links (and their order property) has part of set theory like this:

<img width="869" alt="image" src="https://user-images.githubusercontent.com/7252775/229898323-af0a682d-ce04-4260-8599-d1a98ddc3b62.png">

Now `Max Musterman` has decided that he prefers `Dortmund` as his `HomeAddress`, so he puts the `preference` to `15`. Now we have also changed the ordering of `Address`, which is unintended (only because I changed my preference for where I live, I don't want to change the preference where someone contacts me.)

To summarize, we have a partially ordered set and, inherited types are essentially subsets into the set. When we reorder in that subset (with the same key) we cannot guarantee that we change the order in the superset.

---

So what if we say: you can use `orderBy`, but if you do, it needs to point to a property defined on the current entity type and cannot be inherited. This limits reuse, but ensures that the subset problem described above does not happen. (if I order, I need to order by something different).

Reviewer 3: I think we should avoid any form of restrictions like "it can't be inherited" as that requires us to resolve the full inheritance tree, and as expressed elsewhere in the RFC, implementations would need to be careful about handling external types which don't abide by the spec (potentially accidentally)

Reviewer 5: didn't really get what 'buckets' is from the thread, is it this? being explicit about which type in the link's type chain an order is in relation to, e.g.
```json
{
  "$id": "https://example.com/types/entity-type/work-address/v/1",
  "leftEntityId": "Max",
  "rightEntityId": "Dortmund",
  "leftToRightOrder": {
     "https://example.com/types/entity-type/work-address/v/1": 15,
     "https://example.com/types/entity-type/address/v/1": 30
  }
}
```
you would need to figure out what to do for types that weren't present in this object and whether they inherited a position or not, and whoever was creating these things would need to figure out from context which order(s) should be set

Reviewer 3: Interesting suggestion as well. I could see this being a royal pain to validate though..

It adds a dependency of the _link entity_ onto the _source entity type_ which isn't ideal. If the source entity was updated (or had its entity type updated) then you could invalidate outgoing link entities.. but I guess as part of validation we'd need to check outgoing link entities anyway as some of them may be required.

> you would need to figure out what to do for types that weren't present in this object 

I personally would treat that as a validation error for the link.

Reviewer 3: At the moment I think that's the best solution of the ones we've come up with

Reviewer 3: > didn't really get what 'buckets' is from the thread, is it this

Apologies @CiaranMn I actually missed that this was a question, I thought it was an alternative proposal. The "buckets" thing I was saying would be on the source entity type, and rather than something like this:
```json5
{
  ... 
  "links": {
    "http...some-link-type": {
        "type": "array",
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type"
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

You'd maybe have something like this

```json5
{
  ... 
  "links": {
    "collection1": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type", // could also perhaps be a `oneOf` if we wanted mixed collections for some reason
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type" 
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

Which means you could do

```json5
{
  ... 
  "links": {
    "collection1": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type",
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type"
            }
          ]
        }
      },
      ... // more link types
    },
    "collection2": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type", // same link type
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type" // could be some _other_ entity type (different to above)
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

And then when you create a link, it goes in a specific bucket/collection, rather than just being grouped by link _type_. (Therefore having a unique order within that collection).

I'm pretty sure I don't like this suggestion though, it's adding yet another high-level concept, and this one is weirdly implicitly defined within the entity type, rather than being a standalone re-usable thing. I think if you wanted to group links, we'd want it done within the type system and ontology by creating `Group` or `Path` semantics as enabled by the `links/entities` proposal.

Reviewer 3: Basically removing  (currently undefined but implied) guarantees around ordering and putting that onus on the user

Reviewer 5: Thank you for explaining buckets and grouping. Would need to talk about grouping more (probably easier sync) to have thoughts on it.

> > you would need to figure out what to do for types that weren't present in this object

> I personally would treat that as a validation error for the link.

It depends on whether each link was declared as `ordered` on the entity or not, and how we handle validating that generally. If a link was declared `ordered: true` and we treated that as meaning that an ordering property was required (in which direction? either?), I agree we could treat a missing order _for that type_ as being a validation error. I think it's different if you have a mix of ordered and non-ordered in the same inheritance/extension chain on the same entity type.
